tests/unit/drivers/test_recursive_traversal.py

```diff
@@ -0,0 +1,200 @@
+    if current_adjacency < max_adjacency:
+        for i in range(DOCUMENTS_PER_LEVEL):
+            dc = d.matches.add()
+            dc.granularity = current_granularity
+            dc.adjacency = current_adjacency + 1
+            dc.id = i
+            iterate_build(dc, current_granularity, max_granularity, current_adjacency + 1, max_adjacency)
+
+
+def test_only_granularity():
+    docs = build_docs()
+    driver = SliceQL(
+        start=0,
+        end=1,
+        adjacency_range=(0, 0),
+        granularity_range=(0, 2),
+        recur_on=["chunks"]
+    )
+    driver._traverse_apply(docs)
+    assert len(docs) == 1
+    assert len(docs[0].chunks) == 1
+    assert len(docs[0].chunks[0].chunks) == 1
+    assert len(docs[0].chunks[0].chunks[0].matches) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].chunks[0].matches) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].matches) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].matches[0].chunks) == DOCUMENTS_PER_LEVEL
+
+
+def test_only_adjacency():
+    docs = build_docs()
+    driver = SliceQL(
+        start=0,
+        end=1,
+        adjacency_range=(0, 2),
+        granularity_range=(0, 0),
+        recur_on=["matches"]
+    )
+    driver._traverse_apply(docs)
+    assert len(docs) == 1
+    assert len(docs[0].chunks) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].chunks[0].matches) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].matches) == 1
+    assert len(docs[0].matches[0].chunks) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].matches[0].matches) == 1
+    assert len(docs[0].matches[0].matches[0].chunks) == DOCUMENTS_PER_LEVEL
+
+
+def test_adjacency_chunks():
+    """
+    This combination of `range` and `recur_on` will only `_apply_all` to the root node.
+    Thus, the remaining document structure should still be intact.
+    In practice this combination should not be chosen, anyhow it nicely demonstrates the behaviour.
+    """
+    docs = build_docs()
+    driver = SliceQL(
+        start=0,
+        end=1,
+        adjacency_range=(0, 1),
+        granularity_range=(0, 0),
+        recur_on=["chunks"]
+    )
+    driver._traverse_apply(docs)
+    assert len(docs) == 1
+    assert len(docs[0].chunks) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].chunks[0].matches) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].matches) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].matches[0].chunks) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].matches[0].matches) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].matches[0].matches[0].chunks) == DOCUMENTS_PER_LEVEL
+
+
+def test_granularity_matches():
+    """
+    This combination of `range` and `recur_on` will only `_apply_all` to the root node.
+    Thus, the remaining document structure should still be intact.
+    In practice this combination should not be chosen, anyhow it nicely demonstrates the behaviour.
+    """
+    docs = build_docs()
+    driver = SliceQL(
+        start=0,
+        end=1,
+        adjacency_range=(0, 0),
```

Comment:
Shall we at least mark the adjaceny_range should start from `1`? `adjacency_range=0` means apply should work on `chunks`.